src/trie/merkle_tree.rs

```diff
@@ -395,54 +537,47 @@ impl<H: StarkHash + Send + Sync> MerkleTree<H> {
             ))? {
             Unresolved(hash) => {
                 if path.0.is_empty() {
-                    updates.insert(
+                    updates.push((
                         TrieKey::new(&self.identifier, TrieKeyType::Trie, &[]),
                         InsertOrRemove::Insert(Node::Unresolved(hash).encode()),
-                    );
+                    ));
                     Ok(hash)
                 } else {
                     Ok(hash)
                 }
             }
             Binary(mut binary) => {
                 let left_path = path.new_with_direction(Direction::Left);
-                let left_hash = self.commit_subtree::<DB>(updates, binary.left, left_path)?;
+                let left_hash =
+                    self.commit_subtree::<DB>(updates, binary.left, left_path, hashes)?;
                 let right_path = path.new_with_direction(Direction::Right);
-                let right_hash = self.commit_subtree::<DB>(updates, binary.right, right_path)?;
-                let hash = H::hash(&left_hash, &right_hash);
+                let right_hash =
+                    self.commit_subtree::<DB>(updates, binary.right, right_path, hashes)?;
+                let hash = hashes.next().expect("mismatched hash state");
```

Comment:
Can you explain what is expected from the caller in order for it not to panic?
Like "hashes should be an iterator that ..."
And maybe add a `hashes` section to the doc arguments explaining that it is the caller's responsibility to provide `hashes` such as xyz otherwise it will panic

Reply:
I will do that asap. Thanks :p